### PR TITLE
[eas-cli] add command `eas channel:rollout`

### LINKED ISSUE #685: `eas submit --platform android` results in `CombinedError: [Network] Request Entity Too Large`
### Summary

Running `eas submit --platform android` results in `CombinedError: [Network] Request Entity Too Large` with no further information. See terminal output:

```
$ eas submit --platform android 
✔ Linked to project @owencm/freelancerprojectcoordination (​https://expo.dev/accounts/owencm/projects/freelancerprojectcoordination​)

✔ What would you like to submit? › Latest finished build from EAS
Project ID:                     7ee497ba-f0d0-465f-b723-695e262ca90e
Android package:                com.builtwithsutro.freelancerprojectcoordination
Release track:                  INTERNAL
Changes not sent for a review:  false
Release status:                 DRAFT
Google Service Key:             /Users/owencampbell-moore/Downloads/Sutro/Android-build-test/freelancerprojectcoordination-b9575cb8ba2f49daadeb0b74014ed7b7-signed.aab
Build:                          
    Build ID    :  f2501084-fc06-4adb-8720-d78a22e3742b
    Build Date  :  10/5/2021, 6:10:44 PM
    App Version :  1.1.0
    Version code:  2

✔ Scheduled Android submission
    CombinedError: [Network] Request Entity Too Large
```

Note the error did not occur when picking to upload a build generated by `expo build:android`.

### Environment

```
  Expo CLI 4.12.1 environment info:
    System:
      OS: macOS 10.15.7
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 16.9.1 - /usr/local/bin/node
      Yarn: 1.22.10 - /usr/local/bin/yarn
      npm: 7.21.1 - /usr/local/bin/npm
      Watchman: 2021.09.13.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.4, DriverKit 20.2, macOS 11.1
    IDEs:
      Android Studio: 4.2 AI-202.7660.26.42.7322048
      Xcode: 12.4/12D4e - /usr/bin/xcodebuild
    npmPackages:
      expo: ^41.0.1 => 41.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-41.0.0.tar.gz => 0.63.2 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.12.1
    Expo Workflow: managed
```

### Please specify your device/emulator/simulator platform, model and version

N/A

### Error output

`CombinedError: [Network] Request Entity Too Large`

### Reproducible demo or steps to reproduce from a blank project

I built my project with eas via `eas build --platform android`, and then ran `eas submit --platform android`.

Error occurs whether I pick `› Latest finished build from EAS`, provide the URL to the build or download the build and pick it locally.

### PR DESCRIPTION
**Pull request by jkhales**
# Why

Create a command to manage rollout on a channel.

closes #684
closes #685
closes #686

# How

There are three possible branches:

1. if there is no active rollout, a new one is started based on the `branch` and `percent` specified by the user.
2. if there is an active rollout
  a. user can modify it by changing the percent
  b. user can end it by setting the --end flag and specifying a branch to consolidate to.

# Note

~~There is a bit of of confusing nuance. In a rollout a user is converted to a number in the unit interval and sent to the new Branch if they are less than the rollout percent, old branch otherwis.~~

~~If we use a strict inequality then we get `[0,rolloutPercent)` a weak inequality, `[0,rolloutPercent]`.~~

~~When the rollout percent is `0`, then the strict inequality sends 0 to the new branch since `[0,0)` is the empty set. However in the weak case, `[0,0]` will send a very small number of users to to the new branch.~~

~~Conversely when the rollout percent is `100` we get the opposite confusion.~~

~~Since the new branch is, in general, the more dangerous one, we stick with the strict inequality `[0,rolloutPercent)` so that the a rollout percent of `0` matches user expectations while `100` does not quite.~~

~~We can perhaps treat this server side by continuing to use a strict inequality, but also passing users mapped to 100 on the edge case where the rollout percent is 100 as well.~~

fixed server side: https://github.com/expo/universe/pull/7242

# Test Plan

Manually started/managed/ended a rollout:

Screen Shot 2021-03-26 at 10 34 11 AM

### CODE REVIEW

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    if (isRollout && branchName && !endFlag) {
+      throw new Error(
+        `There is a rollout in progress. You can only either edit the rollout percent or 'end' it.`
+      );
+    }
+
+    let newChannelInfo, logMessage;
+    if (!isRollout) {
+      // start a rollout
+      // this will likely be removed as more commands fill out
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:125]: delete.

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    });
+    const { branchMapping: currentBranchMapping, isRollout } = getBranchMapping(
+      getUpdateChannelByNameForAppResult
+    );
+
+    if (currentBranchMapping.data.length === 0) {
+      throw new Error('The channel is not pointing at any branches.');
+    }
+    if (currentBranchMapping.data.length > 2) {
+      throw new Error('This CLI cannot handle branch mappings with more than 2 branches.');
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:113]:
```suggestion
      throw new Error('"channel:rollout" cannot handle branch mappings with more than 2 branches.');
```

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    if (isRollout && branchName && !endFlag) {
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:116]: add comment that the user is likely trying to start a rollout here since they have supplied a branch name and no end flag.

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      }
+
+      logMessage = `️Started a rollout of branch ${chalk.bold(
+        branchName
+      )} onto channel ${chalk.bold(channelName!)}! ${chalk.bold(
+        percent
+      )}% of users will be directed to ${chalk.bold(branchName)}, ${chalk.bold(
+        100 - percent!
+      )}% to ${oldBranch.name}.`;
+    } else if (!endFlag) {
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:186]: add comment that this is for managing.

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      if (jsonFlag && !percent) {
+        throw new Error(
+          'A rollout is already in progress. You can either end it with the --end flag or modify it with the --percent flag.'
+        );
+      }
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:191]: delet in favor of: https://github.com/expo/eas-cli/pull/290/files#diff-5e715a56413f58e4e51aad67353e66469536e961c39884dd972b572a3d847f2dR206

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        branch => branch.id === newBranchId
+      )[0];
+      const oldBranch = getUpdateChannelByNameForAppResult.app?.byId.updateChannelByName.updateBranches.filter(
+        branch => branch.id === oldBranchId
+      )[0];
+      if (!newBranch || !oldBranch) {
+        throw new Error(
+          `Branch mapping rollout is missing a branch for channel "${channelName}" on app "@${accountName}/${slug}"`
+        );
+      }
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:249]: pull this out into seperate function and use in the `--end` code block as well.

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        channelId: getUpdateChannelByNameForAppResult.app?.byId.updateChannelByName.id!,
+        branchMapping: JSON.stringify(newBranchMapping),
+      });
+      logMessage = `️Rollout on channel ${chalk.bold(
+        channelName
+      )} ended. All traffic is now sent to ${chalk.bold(
+        endOnNewBranch ? newBranch.name : oldBranch.name
+      )}`;
+    } else {
+      throw new Error('');
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:285]: Add message.

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+      throw new Error('This CLI cannot handle branch mappings with more than 2 branches.');
+    }
+
+    if (isRollout && branchName && !endFlag) {
+      throw new Error(
+        `There is a rollout in progress. You can only either edit the rollout percent or 'end' it.`
+      );
+    }
+
+    let newChannelInfo, logMessage;
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:122]: Add comment explaining this big if/else block

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+import {
+  findProjectRootAsync,
+  getBranchByNameAsync,
+  getProjectAccountNameAsync,
+} from '../../project/projectUtils';
+import { promptAsync, selectAsync } from '../../prompts';
+import { updateChannelBranchMappingAsync } from './edit';
+import { getBranchMapping, getUpdateChannelByNameForAppAsync } from './view';
+
+async function getPercentAsync({
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:16]:
```suggestion
async function getRolloutPercentAsync({
```

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    message: promptMessage,
+    initial: 0,
+    validate: (percent: string): true | string => {
+      /**
+       * Don't allow 100 to emphasise to the developer that rollouts have an open right hand endpoint: [0,N)
+       *
+       * This is unavoidable since the comparision must either be a strict inequality or a weak one. A weak
+       * comparison would result in (0,100]. Assuming the new branch is more dangerous, we choose [0,N) to allow
+       * '0 percent' rollouts.
+       */
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/rollout.ts:38]: 🤔

**Comment by jonsamp** [comment]
For the `--end` flag, do we have to specify the branch? I assume not, since channel `main` can only ever have one rollout alive at once.

As a UX thing, can we change text like this:

```
90% of users will be directed to **test**, 10% to **main**.
```

to:

```
90% of users will be directed to the **test** branch; 10% to the **main** branch.
```

Since channels and branches will often have the same name, I think it's important we try to say "branch" and "channel".

File: packages/eas-cli/src/commands/channel/rollout.ts [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+          switch (branch.id) {
+            case newBranch.id:
+              endOnNewBranch = true;
+              break;
+            case oldBranch.id:
+              endOnNewBranch = false;
+              break;
+            default:
+              throw new Error(
+                `The branch "${branchName}"specified by --branch must be one of the branches involved in the rollout: "${newBranch.name}" or "${oldBranch.name}".`
```
Comment (jonsamp) [inline packages/eas-cli/src/commands/channel/rollout.ts:269]:
```suggestion
                `The branch "${branchName}" specified by --branch must be one of the branches involved in the rollout: "${newBranch.name}" or "${oldBranch.name}".`
```

**Review by jonsamp — APPROVED** [review]
This code looks reasonable to me. I read through it all and it makes sense. Happy to do testing if you want to feel more confident about shipping this.

**Review by fson — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/commands/channel/view.ts [hunk 12]
```diff
@@ -62,7 +62,7 @@ function getBranchMapping(
       if (branchMapping.data[0].branchMappingLogic.branchMappingOperator !== 'hash_lt') {
         throw new Error('Branch mapping operator of initial branch mapping must be "hash_lt"');
       }
-      if (!rolloutPercent) {
+      if (rolloutPercent === undefined) {
```
[the remark this answers is not in the document]
Comment (jkhales) [reply packages/eas-cli/src/commands/channel/view.ts:65]: good idea

**Review by fson — APPROVED** [review]
